src/components/lib/posix/posix.c

```diff
@@ -293,6 +298,152 @@ cos_clone(int (*func)(void *), void *stack, int flags, void *arg, pid_t *ptid, v
 	return thd->thdid;
 }
 
+#define FUTEX_WAIT		0
+#define FUTEX_WAKE		1
+#define FUTEX_FD		2
+#define FUTEX_REQUEUE		3
+#define FUTEX_CMP_REQUEUE	4
+#define FUTEX_WAKE_OP		5
+#define FUTEX_LOCK_PI		6
+#define FUTEX_UNLOCK_PI		7
+#define FUTEX_TRYLOCK_PI	8
+#define FUTEX_WAIT_BITSET	9
+
+#define FUTEX_PRIVATE 128
+
+#define FUTEX_CLOCK_REALTIME 256
+
+struct futex_data
+{
+	int *uaddr;
+	struct ps_list_head waiters;
+};
+
+struct futex_waiter
+{
+	thdid_t thdid;
+	struct ps_list list;
+};
+
+#define FUTEX_COUNT 20
+struct futex_data futexes[FUTEX_COUNT];
+
+struct futex_data *
+lookup_futex(int* uaddr)
+{
+	int last_free = -1;
+	int i;
+
+	for (i = 0; i < FUTEX_COUNT; i++) {
+		if (futexes[i].uaddr == uaddr) {
+			return &futexes[i];
+		} else if (futexes[i].uaddr == 0) {
+			last_free = i;
+		}
+	}
+	if (last_free >= 0) {
+		futexes[last_free] = (struct futex_data) {
+			.uaddr = uaddr
+		};
+		ps_list_head_init(&futexes[last_free].waiters);
+		return &futexes[last_free];
+	}
+	printc("Out of futex ids!");
+	assert(0);
+}
+
+/* TODO: Cleanup empty futexes */
+
+struct sl_lock futex_lock = SL_LOCK_STATIC_INIT();
+
+int
```

Comment:
Comment required:

```
/* 
 * precondition: futex_lock is taken
 * Note that this function releases and re-takes that lock, thus critical section state must be re-checked after this returns.
 */
```

If you can add an assertion to check that we have the lock in this function, that is better than the `precondition` above.

I'm aware that none of this matters right now.  It might in the future if the futex implementation evolves, thus this comment.

Reply:
@gparmer It already will assert, although that is subtle. (When we try and release the lock, there is an assert that we hold it.) Will add a comment.